app/novelty/compute_novelty.py

```diff
@@ -3,95 +3,155 @@
+1. Distinguish whether the result is curated or not.
+2. If not curated, calculate recency using all associated publications with the result
+3. If result for the query is a chemical: Molecular Distinctiveness + Clinical Trials/Approval
+4. If result for the query is a gene: Gene Distinctiveness + TDLs
 """
 
+def find_nearest_neighbors( unknown_smiles_dict, known_smiles_dict, similarity_cutoff, num_neighbors):
+    start = time.time()
+    unknown_smiles = {
+        key: value
+        for key, value in unknown_smiles_dict.items()
+        if value != "No SMILES could be found"
+    }
+    known_smiles = {
+        key: value
+        for key, value in known_smiles_dict.items()
+        if value != "No SMILES could be found"
+    }
+
+    # Convert input SMILES to a molecule
+    known_mols = {}
+    for key, value in known_smiles.items():
+        known_mol = Chem.MolFromSmiles(value)
+        if known_mol is not None:
+            known_mols.update({key: known_mol})
+    nearest_neighbor_mapping = {}
+    for unknownkey, value in unknown_smiles.items():
+        query_mol = Chem.MolFromSmiles(value)
+        if query_mol is None:
+            neighbors = []
+            neighbors.append((0, 1))
+            nearest_neighbor_mapping.update({unknownkey: neighbors})
+        else:
+            # Calculate fingerprints for the query molecule
+            query_fp = AllChem.GetMorganFingerprintAsBitVect(query_mol, 2, nBits=2048)
+
+            # Calculate similarity scores between the query molecule and all molecules in the dataset
+            similarities = []
+            for key, mol in known_mols.items():
+                fp = AllChem.GetMorganFingerprintAsBitVect(mol, 2, nBits=2048)
+                similarity = DataStructs.TanimotoSimilarity(query_fp, fp)
+                similarities.append((key, similarity))
+
+            # Sort the similarities in descending order
+            similarities.sort(key=lambda x: x[1], reverse=True)
+
+            # Retrieve the nearest neighbors
+            neighbors = []
+            for i in range(min(num_neighbors, len(similarities))):
+                index, similarity = similarities[i]
+                if similarity >= similarity_cutoff:
+                    neighbors.append((index, similarity))
+            nearest_neighbor_mapping.update({unknownkey: neighbors})
+    end = time.time()
+    print(f"find_nearest_neighbors function time: {end-start}")
+    return nearest_neighbor_mapping
+
+async def mol_to_smile_molpro(molecules):
+    start=time.time()
+    url = "https://molepro-db-transformer.transltr.io/moleprodb/elements/transform"
+    smiles = {}
+    data_mol = list(set(molecules))
+    message = {
+        "controls":
+            [
+            ]
+    }
+    try:
+        for i in data_mol:
+            message["controls"].append({
+                "name": "id",
+                "value": i
+            })
+        response = requests.post(url, json=message)
```

Comment:
We cannot use a synchronous API call here. It will block the entire worker so that it can't work on other asynchronous queries. I personally prefer `httpx`.

Reply:
I've altered the code to work asynchronously using httpx instead of using requests. Do you have any suggestions on how to change the relative imports for gene_nmf and dcc?